graylog2-server/src/main/java/org/graylog2/notifications/NotificationSystemEventPublisher.java

```diff
@@ -0,0 +1,166 @@
+    private static final int MAX_QUEUED_NOTIFICATIONS = 10_000;
+
+    private final DBEventDefinitionService dbEventDefinitionService;
+    private final SystemNotificationRenderService systemNotificationRenderService;
+    private final EventProcessorEngine eventProcessorEngine;
+    private final ScheduledExecutorService scheduler;
+    private final Duration shutdownTimeout;
+
+    private final BlockingQueue<Notification> queuedNotifications;
+    private final AtomicReference<ScheduledFuture<?>> shutdownTask = new AtomicReference<>();
+
+    private Thread executionThread;
+
+    @Inject
+    public NotificationSystemEventPublisher(DBEventDefinitionService dbEventDefinitionService,
+                                            SystemNotificationRenderService systemNotificationRenderService,
+                                            EventProcessorEngine eventProcessorEngine,
+                                            @Named("daemonScheduler") ScheduledExecutorService scheduler,
+                                            @Named("shutdown_timeout") int shutDownTimeoutMs) {
+        this.dbEventDefinitionService = dbEventDefinitionService;
+        this.systemNotificationRenderService = systemNotificationRenderService;
+        this.eventProcessorEngine = eventProcessorEngine;
+        this.scheduler = scheduler;
+        this.queuedNotifications = new LinkedBlockingQueue<>(MAX_QUEUED_NOTIFICATIONS);
+
+        this.shutdownTimeout = Duration.ofMillis(shutDownTimeoutMs);
+    }
+
+    /**
+     * <em>Asynchronously</em> publishes a notification as a system event
+     *
+     * @param notification Notification to publish as a system event. <b>Be aware that this method alters the
+     *                     notification in place.</b>
+     * @return true if the notification was accepted for publishing, false if not
+     */
+    public boolean submit(Notification notification) {
+        try {
+            if (!queuedNotifications.offer(notification)) {
+                LOG.error("Unable to submit system notification for publishing as a system event. Current number of " +
+                                "queued notifications: {}. Max queue size: {}", queuedNotifications.size(),
+                        MAX_QUEUED_NOTIFICATIONS);
+                return false;
+            }
+        } catch (Exception e) {
+            LOG.error("Failed submitting notification for publishing as a system event", e);
+            return false;
+        }
+
+        return true;
+    }
+
+    @Override
+    protected void run() throws Exception {
+        while (isRunning() || !queuedNotifications.isEmpty()) {
+
+            // The shutdown task will interrupt the thread to unblock it, if it got stuck. Because the interrupt
+            // exception is handled in the event publishing code, we have to break out of the run loop manually,
+            // which is what we do here.
+            if (shutdownTask.get() != null && shutdownTask.get().isDone()) {
```

Comment:
Although impossible in this code, but in general that's a tiny race condition
```suggestion
            final var task = shutdownTask.get();
            if (task != null && task.isDone()) {
```

Reply:
thanks!